packages/sandbox/src/lambda_function_log_streamer.ts

```diff
@@ -0,0 +1,169 @@
+import { functionOutputKey } from '@aws-amplify/backend-output-schemas';
+import { LogLevel, Printer } from '@aws-amplify/cli-core';
+import { BackendOutputClient } from '@aws-amplify/deployed-backend-client';
+import {
+  BackendIdentifierConversions,
+  TagName,
+} from '@aws-amplify/platform-core';
+import { BackendIdentifier, BackendOutput } from '@aws-amplify/plugin-types';
+import {
+  CloudFormationClient,
+  DescribeStacksCommand,
+} from '@aws-sdk/client-cloudformation';
+import { LambdaClient, ListTagsCommand } from '@aws-sdk/client-lambda';
+import { Arn, ArnFormat } from 'aws-cdk-lib';
+import { CloudWatchLogEventMonitor } from './cloudwatch_logs_monitor.js';
+
+/**
+ * Logs streamer for customer defined lambda functions in a sandbox.
+ */
+export class LambdaFunctionLogStreamer {
+  /**
+   * Creates an instance of LambdaFunctionLogStreamer
+   */
+  constructor(
+    private readonly lambda: LambdaClient,
+    private readonly cfnClient: CloudFormationClient,
+    private readonly logsMonitor: CloudWatchLogEventMonitor,
+    private readonly backendOutputClient: BackendOutputClient,
+    private readonly printer: Printer
+  ) {}
+
+  public setOutputLocation = (outputLocation: string) => {
+    this.logsMonitor.setOutputLocation(outputLocation);
+  };
+
+  /**
+   * Starts watching logs in the given sandbox.
+   * @param sandboxBackendId The sandbox backend identifier.
+   * @param functionNamesToFilter The function names to filter.
+   */
+  public startWatchingLogs = async (
+    sandboxBackendId: BackendIdentifier,
+    functionNamesToFilter?: string[]
+  ) => {
+    const backendOutput: BackendOutput =
+      await this.backendOutputClient.getOutput(sandboxBackendId);
+
+    const definedFunctionsPayload =
+      backendOutput[functionOutputKey]?.payload.definedFunctions;
+    const deployedFunctionNames = definedFunctionsPayload
+      ? (JSON.parse(definedFunctionsPayload) as string[])
+      : [];
```

Comment:
Maybe we need some sort of "filter" prop for `getBackendMetadata` where it can be instructed to only load metadata from certain verticals? So in this case, we could use it to only load function metadata

Reply:
We don't even need the function stack as technically what we are loading is not a `metadata` unless we start adding the "friendly name" in the metadata or outputs section. In that case we can continue to use `getOutput` here.

I also don't see any issues with depending on `backend-output-schemas` here. `backendOutputClient.getOutput` is the right abstraction here.

Reply:
Taking a closer look at this, it looks like we're only using the functionOutputKey from that package. However, `getOutput` should already have typed keys so we can just use the string directly here and we'll get typechecking without having to import the schema package.

Reply:
I will update it to using the key `"AWS::Amplify::Function"` directly, but wouldn't it just add duplicity. Not that we are ever going to change it but I don't understand the drawback of this dependency?

Reply:
It's not a huge deal. In my head, the `deployed-backend-client` is the "entry point" by which all readers should get info about the backend. the `schema` package is meant to be a contract between the `backend-client` package and the places where the output is written (the constructs). Ideally the schema shouldn't be needed directly by consumers of `deployed-backend-client`

Reply:
It makes sense to have one writer, but I believe it's fine to have multiple readers. We already have `client-config` and `model-generator`

Reply:
That's true but they both use `deployed-backend-client` to get backend output. Neither of them goes directly to CFN

Reply:
This is also coming from `deployed-backed-client`
```ts
const backendOutput: BackendOutput = await this.backendOutputClient.getOutput(sandboxBackendId);
```

Same as https://github.com/aws-amplify/amplify-backend/blob/20bf679430a15277a8c50525dfadfe30a8f4cc1d/packages/cli/src/commands/generate/forms/generate_forms_command.ts#L69 and https://github.com/aws-amplify/amplify-backend/blob/20bf679430a15277a8c50525dfadfe30a8f4cc1d/packages/client-config/src/generate_client_config.ts#L52